src/Microsoft.Diagnostics.Monitoring.Egress.AzureStorage/AzureBlobEgressProvider.cs

```diff
@@ -0,0 +1,186 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Azure;
+using Azure.Storage;
+using Azure.Storage.Blobs;
+using Azure.Storage.Blobs.Models;
+using Azure.Storage.Blobs.Specialized;
+using Microsoft.Extensions.Logging;
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Monitoring.Egress.AzureStorage
+{
+    internal class AzureBlobEgressProvider : EgressProvider<AzureBlobEgressProviderOptions, AzureBlobEgressStreamOptions>
+    {
+        public AzureBlobEgressProvider(AzureBlobEgressProviderOptions options, ILogger logger = null)
+            : base(options, logger)
+        {
+        }
+
+        public override async Task<string> EgressAsync(
+            Func<CancellationToken, Task<Stream>> action,
+            string name,
+            AzureBlobEgressStreamOptions streamOptions,
+            CancellationToken token)
+        {
+            LogAndValidateOptions(streamOptions, name);
+
+            try
+            {
+                var containerClient = await GetBlobContainerClientAsync(token);
+
+                BlobClient blobClient = containerClient.GetBlobClient(GetBlobName(name));
+
+                Logger?.LogDebug("Start invoking stream action.");
+                using var stream = await action(token);
+                Logger?.LogDebug("End invoking stream action.");
+
+                // Write blob content, headers, and metadata
+                Logger?.LogDebug("Start uploading to storage with headers and metadata.");
+                await blobClient.UploadAsync(stream, CreateHttpHeaders(streamOptions), streamOptions.Metadata, cancellationToken: token);
+                Logger?.LogDebug("End uploading to storage with headers and metadata.");
+
+                string blobUriString = GetBlobUri(blobClient);
+                Logger?.LogInformation("Uploaded stream to {0}", blobUriString);
+                return blobUriString;
+            }
+            catch (AggregateException ex) when (ex.InnerException is RequestFailedException innerException)
+            {
+                throw new InvalidOperationException(innerException.Message);
+            }
+            catch (RequestFailedException ex)
+            {
+                throw new InvalidOperationException(ex.Message);
```

Comment:
Consider propagating the caught exception as the inner exception

Reply:
Done